### PR TITLE
Refactor and extract common logic from User and TokenContact

### PR DESCRIPTION
**Pull request by yuliaveres**
Got rid of numerous duplications inside User and TokenContact models, 
considered subclassing from common class instead of conforming to protocol due to several limitations.

### CODE REVIEW

File: Token/Models/TokenContact.swift [hunk 1]
```diff
@@ -33,70 +33,40 @@ public class TokenContact: NSObject, JSONDataSerialization, NSCoding {
...
-
-    public var JSONData: Data {
+    
+    override public var asDict: [String: Any] {
         var imageDataString = ""
         if let image = self.avatar, let data = (UIImagePNGRepresentation(image) ?? UIImageJPEGRepresentation(image, 1.0)) {
             imageDataString = data.hexadecimalString
         }
-
+
```
Comment (3lvis) [inline Token/Models/TokenContact.swift]: Make sure to use SwiftFormat, will clean all of these things for your plus will make sure that we keep the project with a consistent style.

https://github.com/bakkenbaeck/iOS-playbook/blob/master/BEST_PRACTICES.md#linting

File: Token/APIClients/idAPIClient.swift [hunk 2]
```diff
@@ -147,8 +147,8 @@ public class IDAPIClient: NSObject, CacheExpiryDefault {
                 case .success(let json, _):
                     guard let userDict = json?.dictionary else { completion(false); return }
 
-                    User.current?.update(avatar: avatar, avatarPath: userDict["avatar"] as! String)
-
+                    TokenUser.current?.update(avatar: avatar, avatarPath: userDict["avatar"] as! String)
+
```
Comment (Elland) [inline Token/APIClients/idAPIClient.swift]: oops :)

File: Token/Controllers/Contacts/FavoritesController.swift [hunk 3]
```diff
@@ -104,7 +104,7 @@ open class FavoritesController: SweetTableController {
         if let address = UserDefaults.standard.string(forKey: self.selectedContactKey) {
             // This doesn't restore a contact if they are not our contact, but a search result
             DispatchQueue.main.asyncAfter(seconds: 0.0) {
-                guard let contact = self.contact(withAddress: address) else { return }
+                guard let contact = self.contact(with: address) else { return }
```
Comment (Elland) [inline Token/Controllers/Contacts/FavoritesController.swift]: 👏

**Comment by 3lvis** [comment]
👏

File: Token/Controllers/Contacts/FavoritesController.swift (hunk 3 above)
Comment (3lvis) [inline Token/Controllers/Contacts/FavoritesController.swift]: 💥

File: Token/Models/TokenUser.swift [hunk 4]
```diff
@@ -0,0 +1,224 @@
+//
+//  TokenUser.swift
+//  Token
+//
+//  Created by Yuliia Veresklia on 24/04/2017.
+//  Copyright © 2017 Bakken&Bæck. All rights reserved.
```
Comment (3lvis) [inline Token/Models/TokenUser.swift]: Could you delete the attribution header?

File: Token/Views/ContactCell.swift [hunk 5]
```diff
@@ -30,11 +30,6 @@ class ContactCell: UITableViewCell {
 
                 if let avatar = contact.avatar {
                     self.avatarImageView.image = avatar
-                } else {
-                    IDAPIClient.shared.downloadAvatar(path: contact.avatarPath) { image in
```
Comment (3lvis) [inline Token/Views/ContactCell.swift]: What's the reason for removing this piece of code?

Comment (3lvis) [inline Token/Views/ContactCell.swift]: Nevermind, had a second look.

File: Token/Models/TokenUser.swift (hunk 4 above)
Comment (yuliaveres) [inline Token/Models/TokenUser.swift]: oh sorry i need to get used to this :/

Comment (Elland) [inline Token/Models/TokenUser.swift]: Not delete, but add the one we use for this project.

Comment (Elland) [inline Token/Models/TokenUser.swift]: swiftformat will delete it for you, normally. But for this project we use one.

Comment (3lvis) [inline Token/Models/TokenUser.swift]: I guess is this one refactor/user-and-token_contact-common-logic

Comment (yuliaveres) [inline Token/Models/TokenUser.swift]: done

**Comment by 3lvis** [comment]
This looks good to me.

**Comment by Elland** [comment]
Same here. Gonna run it now and look for buggie bugs! 🐛

**Comment by Elland** [comment]
I think we can delete `User.swift`, right?

**Comment by yuliaveres** [comment]
isn't it ? :O

**Comment by Elland** [comment]
No, still used some places too. Wanna remove it all? 🚀

**Comment by yuliaveres** [comment]
can't find any occurrence of old User, i was removing all of them

File: Token/Models/TokenUser.swift [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+            }
+            
+            return [
+                Constants.Keys.address        : address,
+                Constants.Keys.paymentAddress : paymentAddress,
+                Constants.Keys.username       : username,
+                Constants.Keys.about          : about,
+                Constants.Keys.location       : location,
+                Constants.Keys.name           : name,
+                Constants.Keys.avatarDataHex  : avatarPath,
```
Comment (Elland) [inline Token/Models/TokenUser.swift]: Should probably be `keys.avatar`. Duplicated key causes a crash. Also missing all the `self`s. 😄 Also, we don't indent inside lines. Checking the guidelines to be sure it's there now. 👍

File: Token/Models/TokenUser.swift [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+}
+
+extension Notification.Name {
+    public static let CurrentUserDidUpdateAvatarNotification = Notification.Name(rawValue: "CurrentUserDidUpdateAvatarNotification")
+    public static let TokenContactDidUpdateAvatarNotification = Notification.Name(rawValue: "TokenContactDidUpdateAvatarNotification")
+}
+
+public class TokenUser: NSObject,  JSONDataSerialization, NSCoding {
+    
+    struct Constants {
```
Comment (Elland) [inline Token/Models/TokenUser.swift]: Hmm, what is the reason for a structure inside a structure?

File: Token/Models/TokenUser.swift [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+                    self.avatar = image
+                }
+            } else {
+                IDAPIClient.shared.downloadAvatar(path: self.avatarPath) { image in
+                    self.avatar = image
+                }
+            }
+        }
+        
+        if isApp {
```
Comment (Elland) [inline Token/Models/TokenUser.swift]: This doesn't look right to me. I think we need those notifications for both apps and contacts, right? I think they're only different notifications between contacts/apps (app is a subset of contact) and the current user.

Comment (yuliaveres) [inline Token/Models/TokenUser.swift]: okay

File: Token/Models/TokenUser.swift (hunk 7 above)
Comment (yuliaveres) [inline Token/Models/TokenUser.swift]: Used to separate different constants types, string keys, or numeric values etc.

Comment (Elland) [inline Token/Models/TokenUser.swift]: But we don't have any, do we? They're all string. A bit early optimisation maybe? 😄

Comment (yuliaveres) [inline Token/Models/TokenUser.swift]: i can remove nested one

Comment (Elland) [inline Token/Models/TokenUser.swift]: Just realised, is there a reason why this isn't an `enum`? 🤔

Comment (yuliaveres) [inline Token/Models/TokenUser.swift]: nope, i think both are ok. 
i used structs previously for already mentioned grouping reasons (type, purpose of constants inside)
which one do u think is preferred ?

Comment (Elland) [inline Token/Models/TokenUser.swift]: I think enums seem more fitting, but I think I'm fine either way, since we are not using exhaustive switches.

**Pull request merged**